packages/create-subscription/src/createSubscription.js

```diff
@@ -0,0 +1,155 @@
+    subscribe: (
+      source: Property,
+      callback: (value: Value | void) => void,
+    ) => Unsubscribe,
+  |}>,
+): React$ComponentType<{
+  children: (value: Value | void) => React$Node,
+  source: Property,
+}> {
+  const {getValue, subscribe} = config;
+
+  warning(
+    typeof getValue === 'function',
+    'Subscription must specify a getValue function',
+  );
+  warning(
+    typeof subscribe === 'function',
+    'Subscription must specify a subscribe function',
+  );
+
+  type Props = {
+    children: (value: Value) => React$Element<any>,
+    source: Property,
+  };
+  type State = {
+    source: Property,
+    unsubscribeContainer: {
+      unsubscribe: Unsubscribe | null,
+    },
+    value: Value | void,
+  };
+
+  // Reference: https://gist.github.com/bvaughn/d569177d70b50b58bff69c3c4a5353f3
+  class Subscription extends React.Component<Props, State> {
+    state: State = {
+      source: this.props.source,
+      unsubscribeContainer: {
+        unsubscribe: null,
+      },
+      value:
+        this.props.source != null ? getValue(this.props.source) : undefined,
+    };
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+      if (nextProps.source !== prevState.source) {
+        return {
+          source: nextProps.source,
+          unsubscribeContainer: {
+            unsubscribe: null,
+          },
+          value:
+            nextProps.source != null ? getValue(nextProps.source) : undefined,
+        };
+      }
+
+      return null;
+    }
+
+    componentDidMount() {
+      this.subscribe();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+      if (this.state.source !== prevState.source) {
+        this.unsubscribe(prevState);
+        this.subscribe();
+      }
+    }
+
+    componentWillUnmount() {
+      this.unsubscribe(this.state);
+    }
+
+    render() {
+      return this.props.children(this.state.value);
+    }
+
+    subscribe() {
+      const {source} = this.state;
+      if (source != null) {
+        const callback = (value: Value | void) => {
+          this.setState(state => {
+            // If the value is the same, skip the unnecessary state update.
+            if (value === state.value) {
+              return null;
+            }
+
+            // If this event belongs to an old or uncommitted data source, ignore it.
+            if (source !== state.source) {
+              return null;
+            }
+
+            return {value};
+          });
+        };
+
+        // Store subscription for later (in case it's needed to unsubscribe).
+        // This is safe to do via mutation since:
+        // 1) It does not impact render.
+        // 2) This method will only be called during the "commit" phase.
+        const unsubscribe = subscribe(source, callback);
+
+        invariant(
+          typeof unsubscribe === 'function',
+          'A subscription should return either an unsubscribe function or false.',
```

Comment:
Do we support returning false? Looks like we expect it to be a function.

Reply:
Ah! Good catch. I forgot to update the wording. Thanks. :)